grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcHealthCheckService.java

```diff
@@ -0,0 +1,286 @@
+            } else {
+                return;
+            }
+        }
+
+        server = cfg.server();
+        server.addListener(new ServerListenerAdapter() {
+            @Override
+            public void serverStarted(Server server) {
+                serverHealth.setHealthy(true);
+            }
+
+            @Override
+            public void serverStopping(Server server) {
+                // NOT_SERVING will be sent to clients by changing a healthiness of a server
+                serverHealth.setHealthy(false);
+                watchers.keySet().forEach(StreamObserver::onCompleted);
+                watchers.clear();
+            }
+        });
+    }
+
+    @VisibleForTesting
+    ServingStatus checkServingStatus(@Nullable String serviceName) {
+        if (!isHealthy()) {
+            return ServingStatus.NOT_SERVING;
+        }
+        if (Strings.isNullOrEmpty(serviceName)) {
+            return ServingStatus.SERVING;
+        }
+        final ListenableHealthChecker listenableHealthChecker = grpcServiceHealthCheckers.get(serviceName);
+        if (listenableHealthChecker == null) {
+            return ServingStatus.SERVICE_UNKNOWN;
+        }
+        if (listenableHealthChecker.isHealthy()) {
+            return ServingStatus.SERVING;
+        }
+        return ServingStatus.NOT_SERVING;
+    }
+
+    @VisibleForTesting
+    void changeServerStatus(boolean isHealthy) {
+        serverHealth.setHealthy(isHealthy);
+    }
+
+    private void addServerHealthUpdateListener(List<HealthCheckUpdateListener> updateListeners) {
+        serverHealth.addListener(healthChecker -> {
+            updateListeners.forEach(updateListener -> {
+                try {
+                    updateListener.healthUpdated(healthChecker.isHealthy());
+                } catch (Throwable t) {
+                    logger.warn("Unexpected exception from HealthCheckUpdateListener.healthUpdated():", t);
+                }
+            });
+        });
+    }
+
+    private void setInternalHealthUpdateListener(
+            HealthCheckUpdateListener healthCheckUpdateListener,
+            Collection<ListenableHealthChecker> listenableHealthCheckers) {
+        listenableHealthCheckers.forEach(lhc -> {
+            lhc.addListener(healthChecker -> {
+                try {
+                    healthCheckUpdateListener.healthUpdated(healthChecker.isHealthy());
+                } catch (Throwable t) {
+                    logger.warn("Unexpected exception from HealthCheckUpdateListener.healthUpdated():", t);
+                }
+            });
+        });
+    }
+
+    private HealthCheckUpdateListener provideInternalHealthUpdateListener() {
+        return isHealthy -> {
+            if (updating) {
+                return;
+            }
+            updating = true;
+            try {
+                for (Entry<StreamObserver<HealthCheckResponse>, Entry<String, ServingStatus>> entry
+                        : watchers.entrySet()) {
```

Comment:
I think this iterates too many times when it's not necessary.
Let's say that there 1000 `StreamObserver`s for service `A` and 1 `StreamObserver` for service `B`.
When the healthiness of `B` is changed we have to iterate all `StreamObserver`s to find out.
So, how about changing the parameter of the `Map` to  `Map<String, Entry<HealthCheckResponse, ServiceStatus>>` which the `String` is a `serviceName`? so that we can just update the `StreamObserver`s that belong to the specific service.

So we might need to change this method like:
```java
private BiConsumer<String, Boolean> provideInternalHealthUpdateListener() {
    return (name, isHealthy) -> {
    ...
}
```
And use it:
```java
final BiConsumer<String, Boolean> healthCheckUpdateListener = ...
for (Entry<String, ListenableHealthChecker> grpcServiceHealthChecker : grpcServiceHealthCheckers.entrySet()) {
    final String serviceName = grpcServiceHealthChecker.getKey();
    final ListenableHealthChecker healthChecker = grpcServiceHealthChecker.getValue();
    healthChecker.addListener(checker -> healthCheckUpdateListener.accept(serviceName, checker.isHealthy()));
}
```

Reply:
Let me check your idea.
please wait for a moment.

Reply:
Oops, we should use `MultiMap` instead of `Map`. 😅 
This is what I actually meant: 
```java
private Consumer<String> provideInternalHealthUpdateListener() {
    return serviceName -> {
        synchronized (watchers) {
            final ServingStatus currentStatus = checkServingStatus(serviceName);
            final HealthCheckResponse healthCheckResponse = getHealthCheckResponse(currentStatus);
            if (serviceName.isEmpty()) { // server
                final Iterable<Entry<StreamObserver<HealthCheckResponse>, ServingStatus>> streamObservers =
                        watchers.values();
                notifyStreamObservers(streamObservers, healthCheckResponse, currentStatus);
                return;
            }

            final Iterable<Entry<StreamObserver<HealthCheckResponse>, ServingStatus>> streamObservers =
                    watchers.get(serviceName);
            notifyStreamObservers(streamObservers, healthCheckResponse, currentStatus);
        }
    };
}

private static void notifyStreamObservers(
        Iterable<Entry<StreamObserver<HealthCheckResponse>, ServingStatus>> streamObservers,
        HealthCheckResponse healthCheckResponse, ServingStatus currentStatus) {
    for (Entry<StreamObserver<HealthCheckResponse>, ServingStatus> streamObserver :
            streamObservers) {
        if (streamObserver.getValue() != currentStatus) {
            streamObserver.getKey().onNext(healthCheckResponse);
            streamObserver.setValue(currentStatus); // Should we introduce a new type for the entry?
        }
    }
}
```

Reply:
@minwoox 
We've adopted this approach.
Please review.